.github/workflows/release.yml

```diff
@@ -0,0 +1,106 @@
+name: Create Release
+
+on:
+  workflow_run:
+    workflows: ["Build"]
+    types:
+      - completed
+    branches:
+      - main
+
+env:
+  MAJOR_VERSION: "1"
+  MINOR_VERSION: "3"
+
+jobs:
+  release:
+    if: ${{ github.event.workflow_run.conclusion == 'success' }}
+    runs-on: ubuntu-latest
+    permissions:
+      contents: write
+
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Download all artifacts
+        uses: actions/download-artifact@v4
+        with:
+          pattern: UART-CL-*
+          path: artifacts
+          run-id: ${{ github.event.workflow_run.id }}
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Zip Linux and MacOS artifacts
+        run: |
+          cd artifacts
+          zip -r UART-CL-linux-x64.zip UART-CL-linux-x64/*
+          zip -r UART-CL-osx-arm64.zip UART-CL-osx-arm64/*
+          cd ..
+
+      - name: Calculate next version
+        id: version
+        run: |
+          # Get the base version
+          BASE_VERSION="${{ env.MAJOR_VERSION }}.${{ env.MINOR_VERSION }}"
+          
+          # Fetch all tags
+          git fetch --tags
+          
+          # Get count of existing tags that match our base version
+          TAG_COUNT=$(git tag -l "v${BASE_VERSION}.*" | wc -l)
+          
+          # Use tag count as patch version
+          PATCH_VERSION="$TAG_COUNT"
+          
+          # Set the full version
+          FULL_VERSION="v${BASE_VERSION}.${PATCH_VERSION}"
+          echo "new_tag=${FULL_VERSION}" >> $GITHUB_OUTPUT
+          
+          # Generate changelog
+          # Get all version tags and sort them, excluding the current version and anything newer
+          PREV_TAG=$(git tag -l "v*.*.*" | sort -V | while read tag; do
+            # Only keep tags that are strictly less than our new version
+            if [ "$(echo -e "${FULL_VERSION}\n${tag}" | sort -V | head -n1)" = "${tag}" ] && [ "${tag}" != "${FULL_VERSION}" ]; then
+              echo "$tag"
+            fi
+          done | tail -n1 || echo "")
```

Comment:
takes all the version tags, sorts it, then filters out those greater than the tag we are pushing to and returns the first result back from that (which will be the previous release). `"$(echo -e "${FULL_VERSION}\n${tag}" | sort -V | head -n1)" = "${tag}"` specifically places the new version (FULL_VERSION) and then the tag we are testing is not greater than it on a new line, such that we can split it with a pipe, which we then sort, then if the head first line is equal to the tag we must have a tag that was greater thatn FULL_VERSION thus we skip it